typescript/sdk/src/core/EvmIcaModule.ts

```diff
@@ -15,33 +29,182 @@ import {
+      this.args.addresses.interchainAccountRouter,
+    );
   }
 
   public async update(
-    _config: InterchainAccountConfig,
+    expectedConfig: InterchainAccountConfig,
+  ): Promise<AnnotatedEV5Transaction[]> {
+    const actualConfig = await this.read();
+
+    const transactions: AnnotatedEV5Transaction[] = [
+      ...(await this.updateRemoteRoutersEnrollment(
+        actualConfig.remoteIcaRouters,
+        expectedConfig.remoteIcaRouters,
+      )),
+      ...proxyAdminUpdateTxs(
+        this.chainId,
+        this.args.addresses.interchainAccountIsm,
+        actualConfig,
+        expectedConfig,
+      ),
+    ];
+
+    return transactions;
+  }
+
+  private async updateRemoteRoutersEnrollment(
+    actualConfig: DerivedIcaRouterConfig['remoteIcaRouters'],
+    expectedConfig: InterchainAccountConfig['remoteIcaRouters'] = {},
+  ): Promise<AnnotatedEV5Transaction[]> {
+    const transactions: AnnotatedEV5Transaction[] = [
+      ...(await this.getEnrollRemoteIcaRoutersTxs(
+        actualConfig,
+        expectedConfig,
+      )),
+      ...(await this.getUnenrollRemoteIcaRoutersTxs(
+        actualConfig,
+        expectedConfig,
+      )),
+    ];
+
+    return transactions;
+  }
+
+  private async getEnrollRemoteIcaRoutersTxs(
+    actualConfig: Readonly<DerivedIcaRouterConfig['remoteIcaRouters']>,
+    expectedConfig: Readonly<InterchainAccountConfig['remoteIcaRouters']> = {},
+  ): Promise<AnnotatedEV5Transaction[]> {
+    const transactions: AnnotatedEV5Transaction[] = [];
+
+    const routesToEnroll = Array.from(
+      difference(
+        new Set(Object.keys(expectedConfig)),
+        new Set(Object.keys(actualConfig)),
+      ),
+    );
+
+    if (routesToEnroll.length === 0) {
+      return transactions;
+    }
+
+    const domainsToEnroll: string[] = [];
+    const remoteDomainIca: string[] = [];
+    const remoteIsm: string[] = [];
+
+    for (const domainId of routesToEnroll) {
+      domainsToEnroll.push(domainId);
+      remoteDomainIca.push(addressToBytes32(expectedConfig[domainId].address));
+      remoteIsm.push(
+        expectedConfig[domainId].interchainSecurityModule
+          ? addressToBytes32(expectedConfig[domainId].interchainSecurityModule!)
+          : ethers.utils.hexZeroPad('0x', 32),
+      );
+    }
+
+    const remoteTransactions: AnnotatedEV5Transaction[] = domainsToEnroll.map(
+      (domainId) => ({
+        annotation: `Enrolling InterchainAccountRouter on domain ${this.domainId} on InterchainAccountRouter at ${expectedConfig[domainId].address} on domain ${domainId}`,
+        chainId: this.multiProvider.getEvmChainId(domainId),
+        to: expectedConfig[domainId].address,
+        data: InterchainAccountRouter__factory.createInterface().encodeFunctionData(
+          'enrollRemoteRouter(uint32,bytes32)',
+          [
+            this.domainId,
+            addressToBytes32(this.args.addresses.interchainAccountRouter),
+          ],
+        ),
+      }),
+    );
+
```

Comment:
imo these remote transactions in the enroll/unenroll methods should move to their own functions so as not to clutter this method. Generally prefer to have this class only really deal with creating transactions for itself

imagine we're deploying to N chains and call core deploy for each in parallel - if I understand it right here we're going to end up with multiple txs for enrolling the same router?

Reply:
yep we are going to end with at least one tx for each chain where we need to register the new router + the tx on the current chain to register the remote routers with the current one. 

When you say
>  imo these remote transactions in the enroll/unenroll methods should move to their own functions so as not to clutter this method. Generally prefer to have this class only really deal with creating transactions for itself

You prefer that this class only generates txs for the current configured chain?

Reply:
> You prefer that this class only generates txs for the current configured chain?
Yup, iirc this is how the warp module works atm too?

Reply:
> Yup, iirc this is how the warp module works atm too?

Yup, but there we have all the related chains defined in the same config file so for each warp token on a chain a new warp module is created to send the update txs. Here we can't do the same because each core definition belongs to one chain. And probably extracting those remote txs to another function might be error prone when someone wants to reuse the core update logic but the remote ica enrollment/unenrollment is somewhere else and might be forgotten

Reply:
I guess if there's absolutely no way else to do it then I will begrudgingly accept this if we do have a path to make this per-chain again in the future

Reply:
What we might do in the future is to maybe group txs per chain outside of the module, as we do for [`warp apply` txs](https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/main/typescript/cli/src/deploy/warp.ts#L485)

Reply:
agree with paul. the modules should handle the configs in a per chain basis. 

happy to approve this PR as is but lets keep that convention

Reply:
yep i agree we can look for a better way to do this later